test/test_specs.py

```diff
@@ -2348,6 +2348,117 @@ def test_to_numpy(self):
             c.to_numpy(td_fail)
 
 
+@pytest.mark.parametrize("spec", OneHotDiscreteTensorSpec(n=4, shape=[3, 4]))
+@pytest.mark.parametrize(
+    "idx",
+    [
+        5,
+        range(10),
+        np.array([[2, 10]]),
+        (slice(None), slice(1, 2), 1),
+        (1, ..., 2, ..., 3),
+        (1, 1, 1, 1),
+    ],  # [:,1:2,1]
+)
+def test_invalid_indices(spec, idx):
+    with pytest.raises(IndexError):
+        spec[idx]
+
+
+@pytest.mark.parametrize("spec_class", [OneHotDiscreteTensorSpec, DiscreteTensorSpec])
+def test_valid_indices(spec_class):
+    empty_spec = spec_class(0)
+    spec = spec_class(n=4, shape=[3, 4])
+    spec_3d = spec_class(n=4, shape=[5, 3, 4])
+    spec_4d = spec_class(n=6, shape=[5, 3, 4, 6])
+    spec_5d = spec_class(n=7, shape=[5, 3, 4, 6, 7])
+
+    # Integers
+    assert spec[1].shape == torch.Size([4])
+    if not isinstance(spec, OneHotDiscreteTensorSpec):
+        assert spec[0, 1].shape == torch.Size([])
+    # Lists
+    assert spec_3d[[1, 2]].shape == torch.Size([2, 3, 4])
+    assert spec[[0]].shape == torch.Size([1, 4])
+    assert spec[[[[0]]]].shape == torch.Size([1, 1, 1, 4])
+    assert spec[[0, 1]].shape == torch.Size([2, 4])
+    assert spec[[[0, 1]]].shape == torch.Size([1, 2, 4])
+    assert spec_3d[[0, 1], [0, 1]].shape == torch.Size([2, 4])
+    assert spec[[[0, 1], [0, 1]]].shape == torch.Size([2, 2, 4])
+    # Tuples
+    assert spec_3d[1, 2].shape == torch.Size([4])
+    assert spec_3d[(1, 2)].shape == torch.Size([4])
+    assert spec_3d[((1, 2))].shape == torch.Size([4])
+    # Ranges
+    assert spec[range(2)].shape == torch.Size([2, 4])
+    # Slices
+    assert spec[:].shape == torch.Size([3, 4])
+    assert spec[10:].shape == torch.Size([0, 4])
+    assert spec[:1].shape == torch.Size([1, 4])
+    assert spec[1:2].shape == torch.Size([1, 4])
+    assert spec[10:1:-1].shape == torch.Size([1, 4])
+    assert spec[-5:-1].shape == torch.Size([2, 4])
+    assert spec_3d[[1, 2], 3:].shape == torch.Size([2, 0, 4])
+    # None (adds a singleton dimension where needed)
+    assert spec[None].shape == torch.Size([1, 3, 4])
+    assert spec[None, :2].shape == torch.Size([1, 2, 4])
+    expected_shape = [1, 0] if isinstance(spec, OneHotDiscreteTensorSpec) else [1]
+    assert empty_spec[None].shape == torch.Size(expected_shape)
+    # Ellipsis
+    expected_shape = [0] if isinstance(spec, OneHotDiscreteTensorSpec) else []
+    assert empty_spec[...].shape == torch.Size(expected_shape)
+    expected_shape = [2, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 2]
+    assert spec[..., :2].shape == torch.Size(expected_shape)
+    expected_shape = (
+        [2, 1, 1, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 2, 1, 1]
+    )
+    assert spec[..., :2, None, None].shape == torch.Size(
+        expected_shape
+    )  # Matches Numpy's behavior. Tensordict would raise "Not enough dimensions".
```

Comment:
Thanks for commenting that, but since we want to fix it in tensordict I would rather move the comments in the PR to make them more ephemeral it that makes sense.

Reply:
Good call!

Writing for the record that a few tests exhibit a different behavior for numpy and tensordict, especially for indices containing None. We have filed the following [Tensordict issue](https://github.com/pytorch-labs/tensordict/issues/355) and sticked to numpy behavior in these cases.